README.rst
=========
BlanketDB
=========


.. image:: https://img.shields.io/pypi/v/blanketdb.svg
        :target: https://pypi.python.org/pypi/blanketdb

.. image:: https://img.shields.io/travis/luphord/blanketdb.svg
        :target: https://travis-ci.org/luphord/blanketdb

.. image:: https://readthedocs.org/projects/blanketdb/badge/?version=latest
        :target: https://blanketdb.readthedocs.io/en/latest/?badge=latest
        :alt: Documentation Status

.. image:: https://img.shields.io/docker/cloud/build/luphord/blanketdb
        :alt: Docker Automated build

BlanketDB_ is a very simple database written in Python based on SQLite.
It is intended for small IoT projects where you need a quick way to
collect and store data from sensors and other devices.
You communicate to BlanketDB using HTTP GET / POST / DELETE requests.
Request and response bodies are usually JSON, but you can also POST
HTML forms directly to BlanketDB.
There is no schema in the database, you simply store objects in buckets.

BlanketDB is free software provided under a MIT license.
Documentation is available at https://blanketdb.readthedocs.io.

Why is it called BlanketDB? Well, a blanket_ is simple, lightweight, portable
and keeps you warm. But if you really want to relax, you'll need a couch_.

.. _BlanketDB: https://github.com/luphord/blanketdb
.. _blanket: https://github.com/luphord/blanketdb
.. _couch: http://couchdb.apache.org/


Features
--------

* GET / POST / DELETE requests to communicate with BlanketDB
* JSON requests / responses
* HTML forms can POST directly to BlanketDB
* Data stored in buckets
* Schemaless
* Query using various parameters to a HTTP GET request
* Data is stored in a single file on the file system which is a SQLite database
* BlanketDB is a single Python file without any dependencies besides the standard library
* No sequrity whatsoever; BlanketDB is completely open to readers and writers (use with care!)

Quickstart
----------

To install BlanketDB, you'll need a Python (>=3.4) installation with pip:

.. code-block:: console

    $ pip install blanketdb

To use BlanketDB as a standalone database (and communicate over HTTP), enter the following command:

.. code-block:: console

    $ python3 -m blanketdb -i localhost -p 8080 -f /path/to/db.sqlite

BlanketDB will now serve its web interface at http://localhost:8080.
You can open this page in your browser to check if everything works.

To use BlanketDB in a Python project, enter the following code:

.. code-block:: python

    from blanketdb import BlanketDB
    db = BlanketDB('/path/to/db.sqlite')

    # you can now use db using its Python API
    db.store_dict(x='test')['id']
    for entry in db:
        print(entry)

    # db is alse a wsgi conforming callable
    # you can use it e.g. with the wsgi reference implementation
    from wsgiref.simple_server import make_server
    httpd = make_server('localhost', 8080, db)
    httpd.serve_forever()

Detailed documentation is available at https://blanketdb.readthedocs.io.

Credits
-------

Main author and project maintainer is luphord_.

This package was prepared with Cookiecutter_ and the `audreyr/cookiecutter-pypackage`_ project template.

.. _luphord: https://github.com/luphord
.. _Cookiecutter: https://github.com/audreyr/cookiecutter
.. _`audreyr/cookiecutter-pypackage`: https://github.com/audreyr/cookiecutter-pypackage


HISTORY.rst
=======
History
=======

0.4.0 (2020-02-26)
------------------

* Start uwsgi using http protocol by default in DOCKERFILE (s.t. standalone use is possible)
* Overwrite CMD in docker-compose file to communicate via uwsgi protocol between nginx and blanketdb container

0.3.4 (2020-02-26)
------------------

* Support Python 3.8

0.3.3 (2019-12-12)
------------------

* Split tests into Python and HTTP API tests
* Added tests that can be executed against an actual HTTP API of `BlanketDB`

0.3.2 (2019-12-04)
------------------

* Release to trigger build on dockerhub

0.3.1 (2019-03-06)
------------------

* Improved clarity with default values

0.3 (2019-03-06)
----------------

* Type annotations for `BlanketDB`
* Python 3.4 is not supported anymore (as it does not know type annotations)

0.2.2 (2019-01-31)
------------------

* setuptools entrypoint for cli
* quickstart documentation
* added logo

0.2.1 (2019-01-24)
------------------

* fix tag confusion

0.2.0 (2019-01-24)
------------------

* Added CLI for starting `BlanketDB` with `wsgiref.simple_server`
* Tests for `BlanketDB` Web API using `webtest`
* Added documentation for usage and Web API

0.1.0 (2019-01-18)
------------------

* First release on PyPI.


tests/__init__.py
# -*- coding: utf-8 -*-

"""Unit test package for blanketdb."""


tests/test_blanketdb.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""Tests for `blanketdb` module."""


import unittest
from datetime import datetime, timedelta
import json

from webtest import TestApp
import requests

from blanketdb import BlanketDB


class ActualApiAdapeter:

    def __init__(self, base_url):
        self.base_url = base_url.rstrip('/') + '/'

    def perform_request(self, method, path, *args, **kwargs):
        response = method(self.base_url + path, *args, **kwargs)
        response.json = response.json()
        return response

    def get(self, path, status=200):
        return self.perform_request(requests.get, path)

    def post(self, path, body=None, status=200):
        return self.perform_request(requests.post, path, body)

    def post_json(self, path, body=None, status=200):
        body = json.dumps(body)
        return self.perform_request(requests.post, path, body)

    def put(self, path, body=None, status=200):
        return self.perform_request(requests.put, path, body)

    def delete(self, path, status=200):
        return self.perform_request(requests.delete, path)


class TestBlanketdb(unittest.TestCase):
    """Tests for `blanketdb` package."""

    def setUp(self):
        """Set up test instance of `BlanketDB`"""
        self.next_date = datetime(2022, 7, 15)
        self.db = BlanketDB(':memory:', lambda: self.next_date)
        do_actual_api_calls = False
        self.app = ActualApiAdapeter(base_url='http://localhost:8080') \
            if do_actual_api_calls \
            else TestApp(self.db)

    def tearDown(self):
        """Tear down test fixtures, if any."""

    def test_storing_from_python(self):
        '''Test storing of data using Python API'''
        data = dict(a=1, b=2)
        stored = self.db.store(data)
        self.assertEqual(data, stored['data'])
        self.assertEqual(data, self.db[stored['id']]['data'])
        self.assertEqual(1, len(list(self.db)))
        self.assertEqual(data, list(self.db)[0]['data'])
        stored2 = self.db.store_dict(x='test')
        self.assertEqual(dict(x='test'), stored2['data'])
        self.assertEqual(dict(x='test'), self.db[stored2['id']]['data'])
        self.assertEqual(2, len(list(self.db)))
        not_in_db_id = abs(stored['id']) + abs(stored2['id']) + 1
        self.assertEqual(None, self.db[not_in_db_id])
        del self.db[stored['id']]
        self.assertEqual(None, self.db[stored['id']])

    def test_query_from_python(self):
        '''Test `BlanketDB.query` method using Python API'''
        for i in range(10):
            resp = self.db.store_dict(bucket='testbucket', number=i)
            if i == 0:
                self.assertEqual(1, resp['id'], 'SQLite id changed unexpected')
            if i == 4:
                after_four = self.next_date
            self.next_date += timedelta(seconds=4)
        self.assertEqual(10, len(list(self.db.query(bucket='testbucket'))))
        self.assertEqual(0, len(list(self.db.query(bucket='testbucket2'))))
        self.assertEqual(10, len(list(self.db.query(since_id=1))))
        self.assertEqual(9, len(list(self.db.query(since_id=2))))
        self.assertEqual(8, len(list(self.db.query(since_id=3))))
        self.assertEqual(2, len(list(self.db.query(before_id=3))))
        self.assertEqual(6, len(list(self.db.query(since=after_four))))
        self.assertEqual(4, len(list(self.db.query(before=after_four))))
        self.assertEqual(2,
                         len(list(self.db.query(before=after_four, limit=2))))
        i = 10
        for entry in self.db.query(newest_first=True):
            self.assertGreater(i, entry['data']['number'])
            i = entry['data']['number']
        i = -1
        for entry in self.db.query(newest_first=False):
            self.assertLess(i, entry['data']['number'])
            i = entry['data']['number']

    def test_delete_from_python(self):
        '''Test `BlanketDB.delete` method using Python API'''
        for i in range(10):
            resp = self.db.store_dict(bucket='testbucket', number=i)
            if i == 0:
                self.assertEqual(1, resp['id'], 'SQLite id changed unexpected')
            if i == 4:
                after_four = self.next_date
            self.next_date += timedelta(seconds=4)
        self.db.delete(since=after_four)
        self.assertEqual(4, len(list(self.db)))
        self.db.delete(before=after_four)
        self.assertEqual(0, len(list(self.db)))
        self.db.store_dict(bucket='testbucket', irrelevant='data')
        self.db.store_dict(bucket='testbucket2', irrelevant='data')
        self.db.delete(bucket='testbucket')
        self.assertEqual(1, len(list(self.db)))
        self.db.delete()
        self.assertEqual(0, len(list(self.db)))

    def test_basic_web_requests(self):
        '''Test basic web requests'''
        resp = self.app.get('/', status=200)
        self.assertEqual(0, len(resp.json['entries']))
        self.assertEqual(0, resp.json['number_of_entries'])
        self.app.get('/_entry/123', status=404)
        self.app.get('/_entry/', status=400)

    def test_method_not_allowed_requests(self):
        '''Test several method/route combinations which are not allowed'''
        self.app.post('/_entry', status=405)
        self.app.post('/_entry/123', status=405)
        self.app.put('/_entry/', status=405)
        self.app.put('/_entry/123', status=405)
        self.app.put('/anything', status=405)

    def test_create_entry_requests(self):
        '''Test entry creation by request'''
        self.app.post('/', status=201)
        self.app.post('/otherbucket', status=201)
        self.app.post('/', dict(a=1), status=201)  # post form
        self.app.post_json('/', dict(a=1), status=201)  # post json
        self.assertEqual(4, len(self.app.get('/', status=200).json['entries']))
        self.assertEqual(1, len(self.app.get('/otherbucket', status=200)
                                        .json['entries']))
        self.assertEqual(3, len(self.app.get('/default', status=200)
                                        .json['entries']))

    def test_query_requests(self):
        '''Test querying'''
        for i in range(10):
            resp = self.app.post('/testbucket', dict(number=i), status=201)
            if i == 0:
                self.assertEqual(1, resp.json['id'],
                                 'SQLite id changed unexpected')
            if i == 4:
                after_four = self.next_date
            self.next_date += timedelta(seconds=4)
        self.assertEqual(10, self.app.get('/testbucket', status=200)
                                     .json['number_of_entries'])
        self.assertEqual(0, self.app.get('/testbucket2', status=200)
                                    .json['number_of_entries'])
        self.assertEqual(10, self.app.get('/testbucket', dict(since_id=1),
                                          status=200)
                                     .json['number_of_entries'])
        self.assertEqual(9, self.app.get('/testbucket', dict(since_id=2),
                                         status=200)
                                    .json['number_of_entries'])
        self.assertEqual(8, self.app.get('/testbucket', dict(since_id=3),
                                         status=200)
                                    .json['number_of_entries'])
        self.assertEqual(2, self.app.get('/testbucket', dict(before_id=3),
                                         status=200)
                                    .json['number_of_entries'])
        self.assertEqual(6, self.app.get('/testbucket', dict(since=after_four),
                                         status=200)
                                    .json['number_of_entries'])
        self.assertEqual(4, self.app.get('/testbucket',
                                         dict(before=after_four),
                                         status=200)
                                    .json['number_of_entries'])
        self.assertEqual(2, self.app.get('/testbucket',
                                         dict(before=after_four, limit=2),
                                         status=200)
                                    .json['number_of_entries'])
        i = 10
        for entry in self.app.get('/', dict(newest_first=True), status=200) \
                             .json['entries']:
            self.assertGreater(i, entry['data']['number'])
            i = entry['data']['number']
        i = -1
        for entry in self.app.get('/', dict(newest_first=False), status=200) \
                             .json['entries']:
            self.assertLess(i, entry['data']['number'])
            i = entry['data']['number']

    def test_delete_requests(self):
        '''Test entry deletion'''
        # delete by id
        resp = self.app.post('/', status=201)
        id_created = resp.json['id']
        self.assertTrue(isinstance(id_created, int))
        self.app.get('/_entry/{0}'.format(id_created), status=200)
        self.app.delete('/_entry/{0}'.format(id_created), status=200)
        self.app.get('/_entry/{0}'.format(id_created), status=404)
        self.app.delete('/_entry/{0}'.format(id_created), status=404)
        # delete by bucket
        for i in range(3):
            self.app.post('/', status=201)
        for i in range(4):
            self.app.post('/otherbucket', status=201)
        self.assertEqual(3, self.app.get('/default', status=200)
                                    .json['number_of_entries'])
        self.app.delete('/default', status=200)
        self.assertEqual(0, self.app.get('/default', status=200)
                                    .json['number_of_entries'])
        self.assertEqual(4, self.app.get('/otherbucket', status=200)
                                    .json['number_of_entries'])
        self.app.delete('/otherbucket', status=200)
        self.assertEqual(0, self.app.get('/otherbucket', status=200)
                                    .json['number_of_entries'])
        self.app.delete('/otherbucket', status=200)  # still 200
        for i in range(3):
            self.app.post('/', status=201)
        for i in range(4):
            self.app.post('/otherbucket', status=201)
        self.assertEqual(7, self.app.get('/', status=200)
                                    .json['number_of_entries'])
        self.app.delete('/', status=200)  # deletes all, not default bucket
        self.assertEqual(0, self.app.get('/', status=200)
                                    .json['number_of_entries'])


tests/test_helper_functions.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""Tests for helper functions"""


import unittest
import json

from datetime import datetime, date, timedelta
from blanketdb import _parse_form, _parse_dt, _serialize_json, _j


def is_close(dt1, dt2, max_diff_sec=10):
    return abs(dt1.timestamp() - dt2.timestamp()) < max_diff_sec


class TestHelperFunctions(unittest.TestCase):
    """Tests for helper functions"""

    def setUp(self):
        """Set up test fixtures, if any."""

    def tearDown(self):
        """Tear down test fixtures, if any."""

    def test_parse_form(self):
        '''Test function for parsing and type converting forms'''
        self.assertEqual(dict(a='b'), _parse_form('a=b'))
        self.assertEqual(dict(a='b', c='d'), _parse_form('a=b&c=d'))
        self.assertEqual(dict(a=1), _parse_form('a=1'))
        self.assertEqual(1.1, _parse_form('a=1.1')['a'])
        self.assertEqual(float, type(_parse_form('a=1.0')['a']))
        self.assertEqual(dict(a=True), _parse_form('a=true'))
        self.assertEqual(dict(a=False), _parse_form('a=false'))

    def test_parse_date(self):
        '''Test function for customized date parsing'''
        self.assertEqual('', _parse_dt(''))
        # this function only parses dates in custom format
        self.assertEqual('2025-02-03', _parse_dt('2025-02-03'))
        self.assertEqual('', _parse_dt(None))
        self.assertEqual(date.today(), _parse_dt('today'))
        self.assertEqual(date.today() - timedelta(days=1),
                         _parse_dt('yesterday'))
        # days
        self.assertTrue(is_close(datetime.now() - timedelta(days=2),
                        _parse_dt('2d')))
        self.assertTrue(is_close(datetime.now() - timedelta(days=2),
                        _parse_dt('2days')))
        self.assertTrue(is_close(datetime.now() - timedelta(days=2),
                        _parse_dt('2 days')))
        self.assertTrue(is_close(datetime.now() - timedelta(days=1),
                        _parse_dt('1day')))
        # hours
        self.assertTrue(is_close(datetime.now() - timedelta(hours=2),
                        _parse_dt('2h')))
        self.assertTrue(is_close(datetime.now() - timedelta(hours=2),
                        _parse_dt('2hours')))
        self.assertTrue(is_close(datetime.now() - timedelta(hours=2),
                        _parse_dt('2 hours')))
        self.assertTrue(is_close(datetime.now() - timedelta(hours=1),
                        _parse_dt('1hour')))
        # minutes
        self.assertTrue(is_close(datetime.now() - timedelta(minutes=2),
                        _parse_dt('2m')))
        self.assertTrue(is_close(datetime.now() - timedelta(minutes=2),
                        _parse_dt('2min')))
        self.assertTrue(is_close(datetime.now() - timedelta(minutes=2),
                        _parse_dt('2 min')))
        # seconds
        self.assertTrue(is_close(datetime.now() - timedelta(seconds=20),
                        _parse_dt('20s')))
        self.assertTrue(is_close(datetime.now() - timedelta(seconds=20),
                        _parse_dt('20sec')))
        self.assertTrue(is_close(datetime.now() - timedelta(seconds=20),
                        _parse_dt('20 sec')))

    def test_serialize_json(self):
        '''Test function for serializing json with dates'''
        self.assertEqual('{}', _serialize_json({}))
        self.assertEqual('"2030-01-02"', _serialize_json(date(2030, 1, 2)))
        self.assertEqual('"2030-01-02T03:04:05"',
                         _serialize_json(datetime(2030, 1, 2, 3, 4, 5)))

    def test_j(self):
        '''Test function for serializing to json and encoding bytes'''
        self.assertEqual(b'{}', _j({}))
        self.assertEqual(b'"2030-01-02"', _j(date(2030, 1, 2)))
        # avoid indent comparison by decoding json again
        self.assertEqual(dict(a=2, b=3),
                         json.loads(_j(a=2, b=3).decode('utf8')))


blanketdb.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

'''A simple HTTP accessible database for IoT projects.'''

__author__ = 'luphord'
__email__ = 'luphord@protonmail.com'
__version__ = '0.4.0'

import json
import sqlite3
import urllib.parse
from datetime import datetime, date, timedelta
from typing import Dict, Union, Any, Callable, Iterable, List, Optional

from typing import TYPE_CHECKING
if TYPE_CHECKING:
    from wsgiref.types import StartResponse  # noqa: F401
DateLike = Union[str, datetime, date]


def _parse_form(form_s: str) -> Dict[str, Any]:
    '''Parse url encoded form and convert numerical types.'''
    d = dict()  # type: Dict[str, Any]
    for k, v in urllib.parse.parse_qsl(form_s):
        try:
            d[k] = int(v)
        except ValueError:
            try:
                d[k] = float(v)
            except ValueError:
                if v.lower() == 'true':
                    d[k] = True
                elif v.lower() == 'false':
                    d[k] = False
                else:
                    d[k] = v
    return d


def _parse_dt(s: Optional[DateLike]) -> DateLike:
    '''Parse string using custom differential date formats like "2 days".'''
    if not s:
        return ''
    if isinstance(s, datetime):
        return s.isoformat(sep=' ')  # datetime format suitable for sqlite
    s = str(s)
    if s.lower() == 'today':
        return date.today()
    elif s.lower() == 'yesterday':
        return date.today() - timedelta(days=1)
    elif s.lower().endswith('days') \
            or s.lower().endswith('day') \
            or s.lower().endswith('d'):
        try:
            days = int(s.split('d')[0])
        except ValueError:
            return ''
        return datetime.now() - timedelta(days=days)
    elif s.lower().endswith('hours') \
            or s.lower().endswith('hour') \
            or s.lower().endswith('h'):
        try:
            hours = int(s.split('h')[0])
        except ValueError:
            return ''
        return datetime.now() - timedelta(hours=hours)
    elif s.lower().endswith('min') or s.lower().endswith('m'):
        try:
            minutes = int(s.split('m')[0])
        except ValueError:
            return ''
        return datetime.now() - timedelta(minutes=minutes)
    elif s.lower().endswith('sec') or s.lower().endswith('s'):
        try:
            seconds = int(s.split('s')[0])
        except ValueError:
            return ''
        return datetime.now() - timedelta(seconds=seconds)
    return s


def _json_default(obj: Union[datetime, date]) -> str:
    if isinstance(obj, datetime) or isinstance(obj, date):
        return obj.isoformat()
    raise TypeError(type(obj))


def _serialize_json(data: Any, indent: Optional[int] = 2) -> str:
    '''Serialize to json supporting dates.'''
    return json.dumps(data, indent=indent, default=_json_default)


def _j(obj_to_serialize: Any = None, **kwargs: Any) -> bytes:
    '''Serialize `obj_to_serialize` or keyword arguments as dict to json
       and encode to bytes.'''
    if obj_to_serialize is None:
        obj_to_serialize = kwargs
    return _serialize_json(obj_to_serialize).encode('utf8')


_HTTP_STATUS_CODES = {
    200: '200 OK',
    201: '201 Created',
    400: '400 Bad Request',
    404: '404 Not found',
    405: '405 Method Not Allowed',
    415: '415 Unsupported Media Type'
}


class BlanketDB:
    '''A simple HTTP accessible database for IoT projects'''

    _SOURCE = 'FROM blanketdb WHERE (? OR bucket=?) AND rowid>=? ' + \
              'AND timestamp>=? AND (? OR rowid<?) AND (? OR timestamp<?)'
    _QUERY = 'SELECT rowid, * ' + _SOURCE + \
             ' ORDER BY (? * rowid) DESC LIMIT ?;'
    _DELETE = 'DELETE ' + _SOURCE + ';'

    def __init__(self,
                 connection_string: str,
                 now: Callable[[], datetime] = datetime.now) -> None:
        '''Initialize `BlanketDB` instance using a `connection_string`
           that can be understood by SQLite. `now` should be a function
           returning the current datetime (or a suitable test replacement).
        '''
        self.connection = sqlite3.connect(connection_string,
                                          detect_types=sqlite3.PARSE_DECLTYPES)
        with self.connection as conn:
            conn.execute('CREATE TABLE IF NOT EXISTS blanketdb ' +
                         '(bucket text, timestamp timestamp, data text);')
        self.now = now

    def store(self, data: Any, bucket: str = 'default') -> Dict[str, Any]:
        '''Serialize `data` to json and store it under `bucket`.'''
        entry_id = None
        bucket = bucket.lower()
        timestamp = self.now()
        with self.connection as conn:
            c = conn.cursor()
            c.execute('INSERT INTO blanketdb VALUES (?, ?, ?);',
                      (bucket, timestamp, _serialize_json(data, indent=None)))
            entry_id = c.lastrowid
        return dict(id=entry_id, bucket=bucket,
                    timestamp=timestamp.isoformat(), data=data)

    def store_dict(self,
                   bucket: str = 'default',
                   **kwargs: Dict[str, Any]) -> Dict[str, Any]:
        '''Serialize key word args to json and store under `bucket`.'''
        return self.store(kwargs, bucket)

    def __getitem__(self, entry_id: int) -> Optional[Dict[str, Any]]:
        '''Get a stored entry by its `entry_id`.
           Return None if no entry exists for that ID.
        '''
        with self.connection as conn:
            c = conn.execute('SELECT rowid, * FROM blanketdb WHERE rowid=?;',
                             (entry_id,))
            res = c.fetchone()
            if res:
                id, bucket, timestamp, data = res
                return dict(id=id, bucket=bucket,
                            timestamp=timestamp, data=json.loads(data))
            else:
                return None

    def query(self, bucket: str = None,
              since_id: Optional[int] = None,
              since: Optional[DateLike] = None,
              before_id: Optional[int] = None,
              before: Optional[DateLike] = None,
              limit: int = -1, newest_first: bool = True) \
            -> Iterable[Dict[str, Any]]:
        '''Query this `BlanketDB` instance using various optional filters.
           `since` and `since_id` are inclusive, `before` and `before` are
           exclusive regarding the specified value.'''
        is_bucket_requested = bool(bucket)
        if is_bucket_requested:
            bucket = bucket.lower() if bucket else ''
        if since_id is None:
            since_id = 0
        since = _parse_dt(since)
        before = _parse_dt(before)
        with self.connection as conn:
            c = conn.execute(BlanketDB._QUERY,
                             (not is_bucket_requested, bucket,
                              since_id, since,
                              not before_id, before_id, not before, before,
                              1 if newest_first else -1, limit))
            for id, bucket, timestamp, data in c.fetchall():
                yield dict(id=id, bucket=bucket,
                           timestamp=timestamp, data=json.loads(data))

    def __iter__(self) -> Iterable[Dict[str, Any]]:
        '''Iterate over all entries stored in this `BlanketDB` instance.'''
        with self.connection as conn:
            c = conn.execute('SELECT rowid, * FROM blanketdb;')
            for id, bucket, timestamp, data in c.fetchall():
                yield dict(id=id, bucket=bucket,
                           timestamp=timestamp, data=json.loads(data))

    def __delitem__(self, entry_id: int) -> None:
        '''Delete an entry by its `entry_id`.'''
        with self.connection as conn:
            conn.execute('DELETE FROM blanketdb WHERE rowid=?;', (entry_id,))

    def delete(self, bucket: str = None,
               since_id: Optional[int] = None,
               since: Optional[DateLike] = None,
               before_id: Optional[int] = None,
               before: Optional[DateLike] = None) \
            -> Any:
        '''Delete entries from this `BlanketDB` instance
           using various filters. `since` and `since_id` are inclusive,
           `before` and `before` are exclusive regarding the specified value.
        '''
        is_bucket_requested = bool(bucket)
        if is_bucket_requested:
            bucket = bucket.lower() if bucket else ''
        if since_id is None:
            since_id = 0
        since = _parse_dt(since)
        before = _parse_dt(before)
        with self.connection as conn:
            conn.execute(BlanketDB._DELETE, (not is_bucket_requested, bucket,
                         since_id, since,
                         not before_id, before_id, not before, before))
            return conn.execute('select changes();').fetchone()[0]

    def __call__(self,
                 env: Dict[str, Any],
                 start_response: 'StartResponse') \
            -> Iterable[bytes]:
        '''WSGI conform callable method.'''
        def start_json_response(status: int,
                                headers: List[Any] = [('Content-Type',
                                                       'application/json')]) \
                -> None:
            start_response(_HTTP_STATUS_CODES[status], headers)

        path = (str(env['PATH_INFO']) or '/').lower()
        method = str(env['REQUEST_METHOD']).upper()
        qs = _parse_form(str(env['QUERY_STRING']))
        try:
            show_meta = qs.get('meta', True)
            since_id = qs.get('since_id', 0)
            since_id = int(since_id) \
                if isinstance(since_id, (int, str)) \
                else None
            before_id = qs.get('before_id', None)
            before_id = int(before_id) \
                if isinstance(before_id, (int, str)) \
                else None
            since = qs.get('since', None)
            assert since is None or isinstance(since, (str, datetime, date))
            since = _parse_dt(since)
            before = qs.get('before', None)
            assert before is None or isinstance(before, (str, datetime, date))
            before = _parse_dt(before)
            limit = int(str(qs.get('limit', -1)))
            newest_first = bool(qs.get('newest_first', True))
        except Exception as e:
            start_json_response(400)
            yield _j(message='An error occured while' +
                             ' parsing the query parameters: ' + str(e),
                     parameters=qs)
            return

        if method == 'GET':
            if path.startswith('/_entry/'):
                try:
                    entry_id = int(path[8:])
                except ValueError:
                    start_json_response(400)
                    yield _j(message='Path does not contain' +
                                     ' a valid integer ID',
                             path=path)
                    return
                entry = self[entry_id]
                if entry:
                    start_json_response(200)
                    yield _j(entry if show_meta else entry['data'])
                else:
                    start_json_response(404)
                    yield _j(message='Entry does not exist', id=entry_id)
            else:
                start_json_response(200)
                bucket = path[1:]  # type: Optional[str]
                if not bucket:
                    bucket = None  # make it a little more explicit
                entries = list(self.query(bucket, since_id, since,
                               before_id, before, limit, newest_first))
                last_id = max((entry['id'] for entry in entries), default=None)
                entries = [entry if show_meta else entry['data']
                           for entry in entries]
                yield _j(bucket_requested=bucket,
                         since_id=since_id,
                         since=since if since else None,
                         before_id=before_id,
                         before=before if before else None,
                         number_of_entries=len(entries),
                         last_id=last_id,
                         limit=limit if limit > -1 else None,
                         newest_first=newest_first,
                         entries=entries)

        elif method == 'POST':
            if path == '/_entry' or path.startswith('/_entry/'):
                start_json_response(405)
                yield _j(message='The HTTP method is not allowed' +
                                 ' for this path',
                         path=path, method=method)
                return
            try:
                request_body_size = int(env.get('CONTENT_LENGTH', 0))
            except ValueError:
                request_body_size = 0
            request_body = env['wsgi.input'].read(request_body_size)
            bucket = 'default' if path == '/' else path[1:]
            content_type = env.get('CONTENT_TYPE', 'application/json').lower()
            if content_type.startswith('application/json'):
                if request_body == b'':
                    data = None
                else:
                    # ToDo: handle parser errors
                    data = json.loads(request_body.decode('utf8'))
            elif content_type.startswith('application/x-www-form-urlencoded'):
                data = _parse_form(request_body.decode('utf8'))
            else:
                start_json_response(415)
                yield _j(message='Supported media types are application/json' +
                                 ' and application/x-www-form-urlencoded',
                         media_type=content_type)
                return
            entry = self.store(data, bucket=bucket)
            start_json_response(201)
            yield _j(entry)

        elif method == 'DELETE':
            if path.startswith('/_entry/'):
                try:
                    entry_id = int(path[8:])
                except ValueError:
                    start_json_response(400)
                    yield _j(message='Path does not contain' +
                                     ' a valid integer ID',
                             path=path)
                    return
                entry = self[entry_id]
                if entry:
                    start_json_response(200)
                    del self[entry_id]
                    yield _j(entry)
                else:
                    start_json_response(404)
                    yield _j(message='Entry does not exist', id=entry_id)
            else:
                bucket = path[1:]
                if not bucket:
                    bucket = None  # make it a little more explicit
                n = self.delete(bucket, since_id, since, before_id, before)
                start_json_response(200)
                yield _j(bucket_requested=bucket,
                         since_id=since_id,
                         since=since if since else None,
                         before_id=before_id,
                         before=before if before else None,
                         number_of_entries_deleted=n)

        else:
            start_json_response(405)
            yield _j(message='The HTTP method is not allowed for this path',
                     path=path, method=method)


def cli() -> None:
    from argparse import ArgumentParser
    parser = ArgumentParser(description='Start a BlanketDB instance ' +
                                        'using wsgiref.simple_server.')
    parser.add_argument('-i', '--interface', help='Interface to listen on',
                        default='localhost', type=str)
    parser.add_argument('-p', '--port', help='Port to listen on',
                        default=8080, type=int)
    parser.add_argument('-f', '--file', help='Database file to use',
                        default='db.sqlite', type=str)
    args = parser.parse_args()
    from wsgiref.simple_server import make_server
    msg = 'Starting BlanketDB at http://{interface}:{port} using ' + \
          'database file "{file}"'
    print(msg.format_map(vars(args)))
    httpd = make_server(args.interface, args.port, BlanketDB(args.file))
    httpd.serve_forever()


if __name__ == '__main__':
    cli()
